transports/wasm-ext/src/webtransport/connection.rs

```diff
@@ -0,0 +1,177 @@
+use js_sys::{Array, Uint8Array};
+use libp2p_core::multihash::Multihash;
+use libp2p_core::muxing::{StreamMuxer, StreamMuxerEvent, StreamMuxerExt};
+use libp2p_core::{OutboundUpgrade, UpgradeInfo};
+use libp2p_identity::{Keypair, PeerId};
+use send_wrapper::SendWrapper;
+use std::future::poll_fn;
+use std::pin::Pin;
+use std::task::{ready, Context, Poll};
+use wasm_bindgen::JsValue;
+use wasm_bindgen_futures::JsFuture;
+use web_sys::{
+    ReadableStreamDefaultReader, WebTransport, WebTransportBidirectionalStream, WebTransportHash,
+    WebTransportOptions,
+};
+
+use crate::webtransport::cached_js_promise::CachedJsPromise;
+use crate::webtransport::utils::{parse_reader_response, to_js_type, WebTransportEndpoint};
+use crate::webtransport::{Error, Substream};
+
+pub struct Connection {
+    session: SendWrapper<WebTransport>,
+    create_substream_promise: CachedJsPromise,
+    incoming_substream_promise: CachedJsPromise,
+    incoming_substreams_reader: SendWrapper<ReadableStreamDefaultReader>,
+    closed: bool,
+}
+
+impl Connection {
+    pub(crate) fn new(endpoint: &WebTransportEndpoint) -> Result<Self, Error> {
+        let url = format!(
+            "https://{}:{}/.well-known/libp2p-webtransport?type=noise",
+            endpoint.host, endpoint.port
+        );
+
+        let session = if endpoint.certhashes.is_empty() {
```

Comment:
This can't ever be empty with your current parsing strategy of the multiaddr, correct?